### ISSUE TITLE
[Bug] Rendering many Advanced Markers as map children is extremely slow

### ISSUE THREAD

**Issue by sspread** [visgl/react-google-maps#171]
### Description

On initial component render, even a few dozen Advanced Markers as map children will take ~ 10 seconds to initially render map. A few hundred can crash the browser. 

Interestingly, a workaround is to render 1 marker, then set timeout for > 100ms and render many. OR use legacy Marker component.

### Steps to Reproduce

Render 100 AdvancedMarkers as map children (just with key, position props)

### Environment

- Library version: 0.5.0
- Google maps version: weekly (3.55.8)
- Browser and Version: Chrome 120
- OS: Mac

**Comment by sspread** [comment]
Found another workaround, so issue must have something to do with the library loading. Hold the rendering of my component  until marker library is loaded. 

```const Loaded = ({ children }) => {
  const markers = useMapsLibrary('marker')
  const isLoaded = useApiIsLoaded()
  return isLoaded && markers && children
}```

**Comment by usefulthink** [comment]
That's very helpful, thanks. I think I have an Idea what might be going wrong there...

**Comment by usefulthink** [comment]
I tried to somehow reproduce this issue, but even with 1000 advanced markers i can't see any problems. Even in the timeline recording I can only see that creating 1000 markers and adding them to the map takes about 190ms.

**Comment by sspread** [comment]
Are your markers rendered in the first map render? i.e. not added afterwards.

**Comment by usefulthink** [comment]
yep, this is the code i was testing with.

```tsx
const App = () => {
  return (
    <APIProvider apiKey={API_KEY} libraries={['marker']}>
      <Map
        mapId={'bf51a910020fa25a'}
        zoom={3}
        center={{lat: 12, lng: 0}}
        gestureHandling={'greedy'}
        disableDefaultUI={true}>
        {Object.keys(Array.from({length: 1000}))
          .map(Number)
          .map(i => (
            <AdvancedMarker
              key={i}
              position={{
                lat: 20 - Math.floor(i / 50) * 4,
                lng: -100 + 4 * (i % 50)
              }}></AdvancedMarker>
          ))}
      </Map>

      <ControlPanel />
    </APIProvider>
  );
};
```

**Comment by usefulthink** [comment]
This is what it looks like in the profiler: 
![image](https://github.com/visgl/react-google-maps/assets/158792/d68366f1-f51f-4cc7-b019-d48fe0d8c541)

**Comment by sspread** [comment]
Tried with your test code and its crashing Chrome unless I lower count to ~100 (so same issue for me)

**Comment by usefulthink** [comment]
Ok, that's really weird. What kind of GPU do you have? Anything different when opening it in an incognito-tab? Or maybe chrome canary? Tried all the options I have and it works every time.

**Comment by sspread** [comment]
When its slow, its re-rendering my component hundreds of times before painting. So, if I added a console.log('render') in your test 'App' component, it would log a lot. I don't think it has anything to do with chrome or gpu .

**Comment by fc-jeremy** [comment]
I am having this same issue with over 100 advanced markers

**Comment by luis-cicada** [comment]
Im not sure if this is the same issue, but I was having issues after rending clusters and trying to play around with the map. The map was presenting a slow performance. This was not happening on loading but after the cluster rendering. 

The solution: 

using `useMemo` for all the cluster improve the performance A LOT!!

```
const markerCluster = useMemo(() => {
    return (
      <>
        {data.map((point, index) => (
          <AdvancedMarker
              position={{ lat: point.latitude, lng: point.longitude }}
              ref={(marker) => setMarkerRef(marker, index.toString())}
            ></AdvancedMarker>
        ))}
      </>
    )
  }, [markers])

  return <>{markerCluster}</>
```

**Comment by usefulthink** [comment]
@luis-cicada in that case you're also missing the `key`-prop for the `AdvancedMarker` components, so react will recreate **all** markers with every change. Can you try just adding the key-prop and remove the `useMemo`?

**Comment by usefulthink** [comment]
@sspread @fc-jeremy I've been thinking about this and how it could be that me and my colleagues are unable to reproduce this. Do you have a way to provide an URL to something where you see the issue?

Which bundler and settings (babel-presets?) are you using? I was thinking that maybe this could be caused by not using native async/await in the bundled code. That would at least somewhat explain why just waiting once for the loaded-promise seems to resolve this issue.

**Comment by luis-cicada** [comment]
@usefulthink I do have `key-prop` but Im using a custom component and even with `key-prop` it was extremly slow

```
<>
      {data.map((point, index) => (
        <BaseMarker key={index} index={index} point={point} setMarkerRef={setMarkerRef} />
      ))}
</>
```

**Comment by bostrom** [comment]
I experienced the same slowness when using the [marker clustering example from the docs](https://github.com/visgl/react-google-maps/tree/main/examples/marker-clustering) out of the box and swapping out markers dynamically. 

Just rendering the AdvancedMarker without any clustering works fairly fine even with a few thousand markers, but adding the [Markers](https://github.com/visgl/react-google-maps/blob/main/examples/marker-clustering/src/app.tsx#L33) clustering component makes the app crash with a "call stack too deep" error when the marker data changes.

**Comment by usefulthink** [comment]
@bostrom interesting observation, but I think this is a separate issue, I gotta have a look into the marker-clustering at some point..

**Comment by andreasremdt** [comment]
I have a similar observation. Initially, rendering markers is okay-ish in terms of performance, but after swapping markers for a while it's quickly getting slower and slower. And I am talking about 10 markers top, not 100 or 1000. After like 10 swaps, it takes ~5 seconds to render and the entire app becomes unusable.

**Comment by cuongdtt** [comment]
Confirm @sspread workaround works for me with >80 Advance Marker

**Comment by RomanBobrovskiy** [comment]
Any updates here? It's still very slow with >1000 of AdvancedMarker elements, though it's working faster with deprecated Marker element

**Comment by usefulthink** [comment]
Sorry, no updates from our side. There are so far no indications as to what could be the reason for this, and there hasn't been an example provided where we can see the problem occurring (I tried reproducing it myself, but I don't see any problems even with 1000 AdvancedMarkers).

**Comment by jogoley** [comment]
@sspread You are an absolute G. I spent 2 full days trying to figure this out and that simple Loaded component did the trick. Thank you!

**Comment by usefulthink** [comment]
@jogoley since you also experienced this problem and we still haven't had any luck pinning it down, which react version are you using, and are you using the `createRoot`-API or still the older `ReactDOM.render`?

**Comment by yashSangai** [comment]
Solution provided by @sspread could be done like this as well:

```
<APIProvider apiKey={process.env.googleMapsApiKey} libraries={["marker"]}>
           {/* ... any components ... */}
</APIProvider>
```

**Comment by usefulthink** [comment]
It seems like there are still people experiencing this Issue and we would like to properly solve this without requiring any of these workarounds. To do this, we need to be able to reproduce it, which we haven't yet been able to do. 

So, if someone subscribed on this thread could help by providing any more Information, that would be amazing.

**Comment by dogofpavlov** [comment]
I've also run into this exact thing. Rendering 1000 AdvancedMarkers on the map alone runs just fine but when put into a clusterer the page locks for about 8 seconds.

**Comment by watchdogtimer** [comment]
Some data on this that might help, since the behavior been hard to reproduce.

The cost isn't Google's marker API, and it isn't marker content complexity. It's that useAdvancedMarker mints a fresh document.createElement('div') per marker whenever the marker has children, and createPortals into it. React treats every portal container as its own root for event delegation, so the first render into each one calls listenToAllSupportedEvents(container) and registers ~90 native event types on that node. One app root's worth of listeners, per marker.

Measured on one page, same google.maps.Map instance, 2,400 markers, blocking via the Long Task API:

| | raw AdvancedMarkerElement | React + @vis.gl |
| -- | -- | -- |
| Mount blocking | 248ms | 1,728ms |
| Teardown blocking | 0ms (32.8ms sync loop) | 102ms |
| Listeners/marker | 9 | 158 |
| Net listeners left behind | ~7,200 | ~364,800 |

And isolating it to the numChildren > 0 fork, same component, same map:

<AdvancedMarker> with children: 139 listeners/marker
<AdvancedMarker> without children: 9 listeners/marker

The cost is flat regardless of what's inside — one child or fifty, same bundle. Which I think explains why this reproduces for people with lots of markers and not in a minimal repro: it scales with marker count, not with content.

Workaround on our side is to stop passing children and set marker.content to a plain DOM node via the marker ref, which took us from 158 to 9 listeners/marker. One gotcha for anyone trying that: Children.count counts false, so {cond && <Thing/>} still creates the portal.

I don't know the codebase well enough to say what the right fix is here, so I'll leave that to you, but happy to run more measurements if useful.

**Comment by watchdogtimer** [comment]
**TL;DR:** this is Google Maps behavior, not a bug in this library. Clickable markers get sorted on every frame of a zoom, and past a few thousand of them that locks up the main thread. `clickable={false}` plus one delegated listener on the map container fixed it for us.

Apologies for the earlier comment/theory. After some promising testing in dev, the previous fix did not pan out in production, so we kept digging

Correcting my earlier comment. It's `gmpClickable`, the property on Google's `AdvancedMarkerElement` that makes a marker respond to clicks. To know which marker is under the pointer, Google sorts the clickable ones every frame of a zoom.

Setting it true is enough on its own. No click or hover handler required. Measured on a real map in a plain console script, no React, 4,500 markers on the map, zoom level 11 to 18. Longest single main-thread block during the zoom (Long Task API), relative to the fastest run, which was in the tens of milliseconds:

| `gmpClickable` | click/hover handlers | worst block |
| --- | --- | --- |
| true | none | ~990x |
| true | yes | ~590x |
| false | yes | 1x |

`useAdvancedMarker` turns the property on based on the handlers you pass:

```js
gmpClickable = clickable !== undefined
  ? clickable
  : Boolean(onClick) || Boolean(onMouseEnter) || Boolean(onMouseLeave)
```

Sensible default, but it means the natural way to write a hoverable marker list quietly makes every marker clickable, with nothing at the call site suggesting a cost.

Treat those counts as illustrative rather than a limit. The console script needed around 4,500 markers to fall over badly, but our real app froze for ~10s at 1,500 markers and ~40s at 3,000, and others here report trouble lower still. I didn't chase down what moves the threshold.

Our fix was `clickable={false}` on the markers plus one delegated click/pointerover listener on the map container, resolving the target with `closest('[data-property-id]')`. Non-clickable markers still get ordinary DOM pointer events, so delegation catches them. Freeze went from tens of seconds to not noticeable.

Given this isn't a bug and more of a limitation of the underlying GMaps library,  a documentation line saying `onClick`/`onMouseEnter`/ `onMouseLeave` set `gmpClickable`, and that it costs per frame at high marker counts, might be helpful to future developers. Previous comments about clustering might be the better approach, but for our UX, we wanted to avoid clustering.
